kernel/src/hil/bus8080.rs

```diff
@@ -13,6 +13,12 @@ pub enum BusWidth {
     Bits16BE,
 }
 
+pub enum BusAddr8080 {
+    BusAddr8([u8; 1]),
+    BusAddr16BE([u8; 2]),
+    BusAddr16LE([u8; 2]),
```

Comment:
I'm not sure how to interpret this. Are the bytes stored in the proper endianess in the associated array? If so, we don't need to know the endianess. If not, why not just use a `u16`?

Reply:
> Are the bytes stored in the proper endianess in the associated array?

I'm _guessing_ that this is for some imaginary hardware that converts addresses from the target's native endianess to what'll be put on the wire (network byte order?).

Alternatively, there may be buses that have little-endian, and others that have big-endian byte orders. But presumably that would be dictated by the bus itself, and not need to be communicated to / dictated by upper layers?

I'm coming around to this idea. We should just use `u{16,32,64}::{to,from}_be_bytes` to generate the on-wire representation.

Reply:
> > Are the bytes stored in the proper endianess in the associated array?
> 
Yes.

> I'm _guessing_ that this is for some imaginary hardware that converts addresses from the target's native endianess to what'll be put on the wire (network byte order?).
> 

No, this is a software problem as the underlying buses are usually serial.

> Alternatively, there may be buses that have little-endian, and others that have big-endian byte orders. But presumably that would be dictated by the bus itself, and not need to be communicated to / dictated by upper layers?

Busses are usually serial (I2C, SPI), meaning they send data byte by byte, without any knowledge about endianness. The receiving device, in out case the screen, is the one dictating the endianness, as it is the one interpreting the data. In the case of Tock, it is the lower-level screen driver that dictates the address endianness. Except for the 8080 bus (which is a parallel one) the endianness is transparent.

> 
> I'm coming around to this idea. We should just use `u{16,32,64}::{to,from}_be_bytes` to generate the on-wire representation.

We do use these functions when we generate the addresses in the implementation of the `From` trait.

Reply:
> I'm not sure how to interpret this. Are the bytes stored in the proper endianess in the associated array? If so, we don't need to know the endianess. If not, why not just use a `u16`?

The bytes are stored in the proper endianness, ready to be sent on a serial bus. Even if the bytes are in the proper order, we still need to know the endianness due to buses, like the 8080, that are serial.

We use an array instead of an `u16` as the conversion is done in the `From` trait. We could use an `u16` and defer the conversion to the `BusAddr::bytes` function, but this implies creating a new array every time this function is called.

Reply:
If we use a serial bus, there is no problem. For buses like 8080, which is parallel, the endianness matters.

Reply:
Ok I think I finally understand this. In my opinion it makes more sense to just store the address as a `BusAddr16(u16)` and then convert from/to whatever byte order the bus is using. But if it is cleaner/simpler to just store the bytes as they came in and what endianness they were when they were stored that works too.

But a comment explaining this would be helpful.

Reply:
I have a preference for `BusAddr16(u16)`.

Reply:
What is target endianness? The address is just the address (stored as a u16). The user of the address just puts the MS byte (`addr >> 8`) wherever it needs and the LS byte (`addr & 0xFF`) wherever it needs.

Reply:
I guess it might help to have some concrete examples to work through here.

@bradjc Your suggestion makes sense when endianness is always determined by the bus itself.

When two peripherals on a bus can have different edianness expectations for their addresses, then encoding it in the type makes sense.

It never makes sense to have it both in the enum variant `{BE,LE}`, _and_ implicitly in an array `[_; u8]`.

@inesmaria08 @alexandruradovici Can you provide some concrete examples here?

Reply:
> When two peripherals on a bus can have different edianness expectations for their addresses, then encoding it in the type makes sense.

That seems like a very odd design. But ok. I would probably argue that half of the addresses are wrong, but, systems can be complicated.

Reply:
Based on what I've read, it seems we tend keep the changes as they are. I'm uncertain about how to proceed from here.

Reply:
I think this just needs documentation. The enum variant encodes what endianness the device uses on the bus. The bytes are then stored somehow, and that somehow should be documented. Is byte index 0 always sent first? Or is byte index 0 always the least significant byte? Or the most significant byte?